entity-service/src/main/kotlin/com/egm/stellio/entity/service/EntityOperationService.kt

```diff
@@ -0,0 +1,116 @@
+ */
+@Component
+class EntityOperationService(
+    private val entityService: EntityService,
+    private val graphBuilder: GraphBuilder
+) {
+
+    /**
+     * Splits [entities] by their existence in the DB.
+     */
+    fun splitEntitiesByExistence(entities: List<ExpandedEntity>): Pair<List<ExpandedEntity>, List<ExpandedEntity>> =
+        entities.partition {
+            entityService.exists(it.id)
+        }
+
+    /**
+     * Creates a batch of [entities].
+     *
+     * @return a [BatchOperationResult]
+     */
+    fun create(entities: List<ExpandedEntity>): BatchOperationResult {
+        val (graph, invalidRelationsErrors) = graphBuilder.build(entities)
+
+        val (naiveBatchOperationResult, entitiesLeft) = createEntitiesWithoutCircularDependencies(graph)
+        val (createSuccess, circularCreateErrors) = createEntitiesWithCircularDependencies(entitiesLeft.toList())
+
+        val success = naiveBatchOperationResult.success.plus(createSuccess)
+        val errors = invalidRelationsErrors.plus(naiveBatchOperationResult.errors).plus(circularCreateErrors)
+
+        return BatchOperationResult(ArrayList(success), ArrayList(errors))
+    }
+
+    private fun createEntitiesWithoutCircularDependencies(graph: Graph<ExpandedEntity>): Pair<BatchOperationResult, MutableSet<ExpandedEntity>> {
+        val batchOperationResult = BatchOperationResult(arrayListOf(), arrayListOf())
+        val mutableGraph = copyOf(graph)
+
+        /**
+         * Gets a list of leaves, with 0 successors.
+         */
+        fun <T> Graph<T>.getLeaves(): List<T> = nodes().filter { outDegree(it) == 0 }
+
+        do {
+            val leaves = mutableGraph.getLeaves()
+
+            val res = leaves.parallelStream().map {
+                try {
+                    entityService.createEntity(it)
+                    Pair(it.id, null)
+                } catch (e: BadRequestDataException) {
+                    Pair(null, BatchEntityError(it.id, arrayListOf(e.message)))
+                }
+            }
+
+            res.toList().fold(batchOperationResult, { result, (entity, error) ->
+                entity?.let { result.success.add(it) }
+                error?.let { result.errors.add(it) }
+                result
+            })
+
+            leaves.forEach {
+                mutableGraph.removeNode(it)
+            }
+        } while (leaves.isNotEmpty())
+
+        return Pair(batchOperationResult, mutableGraph.nodes())
+    }
+
+    /*
+     * Creates given entities into database.
+     * When there are circular dependencies between nodes, the creation should be achieved in 2 steps:
+     * first create a "temp" entity with almost no attributes
+     * then create the relationships on these attributes
+     */
+    private fun createEntitiesWithCircularDependencies(entities: List<ExpandedEntity>): BatchOperationResult {
+        entities.forEach { entity ->
+            entityService.createTempEntityInBatch(entity.id, entity.type, entity.contexts)
```

Comment:
You can move this method in `EntityOperationService`, it is not meant to be used in other contexts

Reply:
I've moved this method in ```EntityOperationService```